lib/hsa/mcwamp_hsa.cpp

```diff
@@ -4587,10 +4577,124 @@ void HSADispatch::overrideAcquireFenceIfNeeded()
 }
 
 inline hsa_status_t
-HSADispatch::setLaunchConfiguration(int dims, size_t *globalDims, size_t *localDims,
-                                 int dynamicGroupSize) {
+HSADispatch::setLaunchConfiguration(const int dims, size_t *globalDims, size_t *localDims,
+                                    const int dynamicGroupSize) {
     assert((0 < dims) && (dims <= 3));
 
+#if KALMAR_DEBUG && HCC_DEBUG_KARG
+    std::cerr << "static group segment size: " << kernel->static_group_segment_size << "\n";
+    std::cerr << "dynamic group segment size: " << dynamicGroupSize << "\n";
+#endif
+    // Set group dims
+    // for each workgroup dimension, make sure it does not exceed the maximum allowable limit
+    const uint16_t* workgroup_max_dim = device->getWorkgroupMaxDim();
+
+    int workgroup_size[3] = { 1, 1, 1};
+
+    // Check whether the user specified a workgroup size
+    if (localDims[0] != 0) {
+      for (int i = 0; i < dims; i++) {
+        // If user specify a group size that exceeds the device limit
+        // throw an error
+        if (localDims[i] > workgroup_max_dim[i]) {
+          std::stringstream msg;
+          msg << "The extent of the tile (" << localDims[i] 
+              << ") exceeds the device limit (" << workgroup_max_dim[i] << ").";
+          throw Kalmar::runtime_exception(msg.str().c_str(), -1);
+        }
+        workgroup_size[i] = localDims[i];
+      }
+    }
+    else {
+
+      constexpr unsigned int recommended_flat_workgroup_size = 64;
+
+      // user didn't specify a workgroup size
+      if (dims == 1) {
+        workgroup_size[0] = recommended_flat_workgroup_size;
+      }
+      else if (dims == 2) {
+
+        // compute the group size for the 1st dimension
+        for (unsigned int i = 1; ; i<<=1) {
+          if (i == recommended_flat_workgroup_size
+              || i >= globalDims[0]) {
+            workgroup_size[0] = i;
```

Comment:
Should we be choosing the minimum between `i`, `globalDims[0]`, and `recommended_flat_workgroup_size` here?

Reply:
yes, we could take a min between i and globalDims[0] which would lead to a more finer grain group size, I also updated the logic for other dimensions due to this change.